### PR TITLE
feat: rewrite to use Deno FFI

### LINKED ISSUE #6: Use webview/webview_rust instead of boscop/web-view
This module will switch to [webview/webview_rust](https://github.com/webview/webview_rust) as soon as possible due to many of [boscop/web-view](https://github.com/Boscop/web-view) problems. This will also make sure to use [webview/webview](https://github.com/webview/webview) code instead of boscop's fork.

### LINKED ISSUE #76: crash: invalid promise id: expected type `rusty_v8::data::Integer`, got `rusty_v8::data::Value`
I am getting a crash with the following error:

```
...
Check https://deno.land/x/webview/examples/local.ts
Download https://github.com/webview/webview_deno/releases/download/0.5.6/libwebview_deno.so
error: Uncaught (in promise) TypeError: invalid promise id: expected type `rusty_v8::data::Integer`, got `rusty_v8::data::Value`
  const response = Plug.core.dispatch(opId, encode(data))!;
                             ^
    at sync (https://deno.land/x/webview@0.5.6/plugin.ts:53:30)
    at new Webview (https://deno.land/x/webview@0.5.6/webview.ts:58:22)
    at https://deno.land/x/webview@0.5.6/examples/local.ts:11:17
```

### LINKED ISSUE #79: Deno 1.10 support
deno 1.10.1 (release, x86_64-pc-windows-msvc)
v8 9.1.269.27
typescript 4.2.2

```
error: TS2345 [ERROR]: Argument of type '{ url: string; }' is not assignable to parameter of type 'boolean | undefined'.
  Type '{ url: string; }' is not assignable to type 'true'.
  { url: `data:text/html,${encodeURIComponent(html)}` },
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/webview@0.6.0-pre.0/examples/local.ts:12:3
```

### LINKED ISSUE #89: Incompatible with latest Deno (v1.11.3)
Hi,

The example, and fix documented in #79 do not work for me in the latest Deno versions (v.1.11.3).

The advise on that issue is to downgrade to v1.8.3, however that is not always an option.

Is there an estimated timeline for a new release compatible with the latest Deno?

John

### LINKED ISSUE #94: Type Error for {url:string;}
TS2345 [ERROR]: Argument of type '{ url: string; }' is not assignable to parameter of type 'boolean | undefined'.
  Type '{ url: string; }' is not assignable to type 'true'.

### LINKED ISSUE #96: Deno.openPlugin is not a function
With the default example code and the latest version of Deno I get:

> error: Uncaught (in promise) TypeError: Deno.openPlugin is not a function
>  return Deno.openPlugin(file.path);

### LINKED ISSUE #98: Main example doesn't work, breaking change in latest build
I just ran this command as recommended from the readme:

```sh
deno run -Ar --unstable https://deno.land/x/webview/examples/local.ts
```

Results in this error:

> Check https://deno.land/x/webview/examples/local.ts
> error: TS2345 [ERROR]: Argument of type '{ url: string; }' is not assignable to parameter of type 'boolean | undefined'.
>   Type '{ url: string; }' is not assignable to type 'true'.
>   { url: `data:text/html,${encodeURIComponent(html)}` },
>   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     at https://deno.land/x/webview@0.6.0-pre.0/examples/local.ts:12:3

### PR DESCRIPTION
**Pull request by littledivy**
## Breaking Changes

* Synchronous API. 
  - Most of the calls get offloaded to the web renderer anyways. Also probably don't want to go through the overhead of spawning a tokio green thread in non blocking FFI.
  - Note that this does ~~not~~ block the JS event loop but spawning **multiple webviews (one at a time) is still possible**. This is a design limitation. One solution to this is native support for [FFI callbacks (WIP)](https://github.com/denoland/deno/pull/13162)
* Leverage JS setters 
  - `setTitle` changes to a setter on `webview.title`. Same for `setSize`

cc @eliassjogreen 

Closes #96 
Closes #98 
Closes #94 
Closes #6 
Closes #76 
Closes #89 
Closes #79

### CODE REVIEW

**Comment by littledivy** [comment]
One final blocker https://github.com/denoland/deno/pull/13162

**Comment by brandonros** [comment]
might want to update the README to disclose that you got rid of the need for https://github.com/webview/webview_deno#deno https://github.com/denosaurs/plug

**Comment by eliassjogreen** [comment]
@brandonros Plug will still be needed for this pr as to allow for versioned caching of dylibs and easy loading from urls.

File: src/ffi.ts
Comment (brandonros) [inline src/ffi.ts:1]: guessing this needs to get changed from Deno.dlopen to `Plug.prepare` instead then? https://github.com/denosaurs/plug

Comment (eliassjogreen) [reply src/ffi.ts:1]: Yep

File: src/webview.ts
Comment (justinmchase) [inline src/webview.ts:23]: Should these be marked `private`? What is the meaning of the `#` here?

Comment (Miaourt) [reply src/webview.ts:23]: It's the native sugar to make a class field private at runtime (with an associated computation cost at runtime)
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

File: examples/remote.ts
Comment (littledivy) [reply examples/remote.ts:5]: Please read the PR description carefully. The API is intentionally synchronous.

File: examples/external.ts
Comment (littledivy) [reply examples/external.ts:19]: `webview.run` is blocked on callback support. https://github.com/webview/webview_deno/pull/99#issuecomment-998724851

File: src/webview.ts
Comment (littledivy) [reply src/webview.ts:23]: yes, these are truly private (private at runtime) fields.

**Comment by littledivy** [comment]
@eliassjogreen Callback support in Deno FFI is blocked on rusty_v8's Locker API and we (Deno) are not so sure about the concept of Rust calling into V8. 
To unblock deno_webview, maybe we should just do a wrapper to the callbacks? This increases maintenance but was already the case with plugins. WDYT?

**Comment by eliassjogreen** [comment]
Sure, that sounds like a good idea until callback support is added.

File: src/webview.ts
Comment (justinmchase) [reply src/webview.ts:23]: giphy

**Comment by nnmrts** [comment]
<a href="https://www.boss.dev/issue/github/I_kwDODp5Jss4_eInx"></a>

Sorry for going a bit off-topic but I just wanted to say I added a small bounty for issue #96 (which should be closed by this PR) to express my excitement and gratefulness. Thanks for the work y'all put in so far. I would try to help if I had any knowledge of rust or this FFI stuff. 😝

File: src/webview.ts
Comment (eliassjogreen) [reply src/webview.ts:19]: I personally discourage use of typescript enums as they are non-es compliant and one of the few features of typescript which actually generate custom, [also for future proofing](https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums).

**Review by eliassjogreen — CHANGES_REQUESTED** [review]

File: src/ffi.ts
Comment (littledivy) [reply src/ffi.ts:13]: Unfortunately ~~(or maybe fortunately)~~, I'm not near a windows machine to test this.

Comment (eliassjogreen) [reply src/ffi.ts]: You are certainly pretty lucky, I will clone it locally and see if i can work it out this weekend

**Comment by brandonros** [comment]
you might want to delete the empty .gitmodules leftover file

**Comment by brandonros** [comment]
@eliassjogreen anything left review wise?

**Comment by brandonros** [comment]
```
error: TS2571 [ERROR]: Object is of type 'unknown'.
    sys.symbols.deno_webview_get_recv().then((recv) => {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    at https://deno.land/x/webview@0.7.0-pre.0/src/webview.ts:62:5

TS7006 [ERROR]: Parameter 'recv' implicitly has an 'any' type.
    sys.symbols.deno_webview_get_recv().then((recv) => {
                                              ~~~~
    at https://deno.land/x/webview@0.7.0-pre.0/src/webview.ts:62:47

Found 2 errors.

```

Not sure if you need to release?

**Comment by eliassjogreen** [comment]
Strange i am not getting the same error when running locally, have you tried the `--reload` flag? could be some cache issue

**Pull request merged**
